src/ReactART.js

```diff
@@ -12,312 +12,71 @@
+  require('art/modes/fast-noSideEffects'),
 );
 
-const Transform = require('art/core/transform');
 const Mode = require('art/modes/current');
-
+const Transform = require('art/core/transform');
+const invariant = require('fbjs/lib/invariant');
 const React = require('react');
-const ReactDOM = require('react-dom');
-const ReactInstanceMap = require('react/lib/ReactInstanceMap');
-const ReactMultiChild = require('react/lib/ReactMultiChild');
-const ReactUpdates = require('react/lib/ReactUpdates');
+const ReactFiberReconciler = require('react-dom/lib/ReactFiberReconciler');
 
-const emptyObject = require('fbjs/lib/emptyObject');
-const invariant = require('fbjs/lib/invariant');
+const { Component } = React;
 
-const assign = require('object-assign');
 const pooledTransform = new Transform();
 
-// Utilities
+const EVENT_TYPES = {
+  onClick: 'click',
+  onMouseMove: 'mousemove',
+  onMouseOver: 'mouseover',
+  onMouseOut: 'mouseout',
+  onMouseUp: 'mouseup',
+  onMouseDown: 'mousedown',
+};
+
+const TYPES = {
+  CLIPPING_RECTANGLE: 'ClippingRectangle',
+  GROUP: 'Group',
+  SHAPE: 'Shape',
+  TEXT: 'Text',
+};
+
+/** UTILITIES */
 
 function childrenAsString(children) {
   if (!children) {
     return '';
-  }
-  if (typeof children === 'string') {
+  } else if (typeof children === 'string') {
     return children;
-  }
-  if (children.length) {
+  } else if (children.length) {
     return children.join('\n');
-  }
-  return '';
-}
-
-function createComponent(name) {
-  const ReactARTComponent = function(element) {
-    this.node = null;
-    this.subscriptions = null;
-    this.listeners = null;
-    this._mountImage = null;
-    this._renderedChildren = null;
-    this.construct(element);
-  };
-  ReactARTComponent.displayName = name;
-  for (let i = 1, l = arguments.length; i < l; i++) {
-    assign(ReactARTComponent.prototype, arguments[i]);
-  }
-
-  return ReactARTComponent;
-}
-
-/**
- * Insert `node` into `parentNode` after `referenceNode`.
- */
-function injectAfter(parentNode, referenceNode, node) {
-  let beforeNode;
-  if (node.parentNode === parentNode &&
-      node.previousSibling === referenceNode) {
-    return;
-  }
-  if (referenceNode == null) {
-    // node is supposed to be first.
-    beforeNode = parentNode.firstChild;
   } else {
-    // node is supposed to be after referenceNode.
-    beforeNode = referenceNode.nextSibling;
-  }
-  if (beforeNode && beforeNode.previousSibling !== node) {
-    // Cases where `node === beforeNode` should get filtered out by earlier
-    // checks and the behavior isn't well-defined.
-    invariant(
-      node !== beforeNode,
-      'ReactART: Can not insert node before itself'
-    );
-    node.injectBefore(beforeNode);
-  } else if (node.parentNode !== parentNode) {
-    node.inject(parentNode);
+    return '';
   }
 }
 
-// ContainerMixin for components that can hold ART nodes
+/** COMPONENTS */
 
-const ContainerMixin = assign({}, ReactMultiChild, {
+class Node {
```

Comment:
You have a wrapper class around every instance. However, they don't seem to have much interesting state associated with them so we're allocating them only to store a pointer to the ART instance.

Could you just use the ART instance as the instance and rely on pattern matching to pick the logic branch instead of OO methods?